### DISCUSSION TITLE
How to enable <tab> autocompletion supported in cmd2

### DISCUSSION THREAD

**Discussion by Sripadvallabh** [python-cmd2/cmd2#1317] [Q&A]
I had enabled tab autocomplete support using completer functions in my cmd2 app:

```
(lӓut-leaf2) parse show ip route
summary           supernets-only    vrf               {protocol}        {route}
(lӓut-leaf2) parse show ip route
```

But when users type \<tab\> all they see are spaces and the autocompletions doesn't show up.

Should we install certain modules or enable some settings for their shell to pick up and recognize the tab to be an autocomplete?

**Comment by kmvanbrunt** [comment]
As long as `stdin` is a terminal and the Python environment has a supported `readline` library installed, then `cmd2's` tab completion is on by default.

Supported `readline` libraries:

1. Linux - readline
2. Windows - pyreadline3 (imports as readline)
3. Mac - gnureadline

What OS and terminal are your users running in?
What version of Python?

If on Linux or Windows, can you `import readline` successfully?
If on Mac, can you `import gnureadline` successfully?

**Comment by Sripadvallabh** [comment]
I have installed gnureadline on mac but still facing the error:
```
𒄊   source Documents/venc/bin/activate
(venc) 𒄊   pip install gnureadline
Collecting gnureadline
  Using cached gnureadline-8.2.10-cp312-cp312-macosx_10_9_x86_64.whl.metadata (11 kB)
Using cached gnureadline-8.2.10-cp312-cp312-macosx_10_9_x86_64.whl (160 kB)
Installing collected packages: gnureadline
Successfully installed gnureadline-8.2.10
(venc) 𒄊
(venc) 𒄊
(venc) 𒄊   cd Documents/live-aut
(venc) 𒄊   python3 laut.py
(lӓut-TB-NONE) testb >>>>>>>>> NOT AUTOCOMPLETING ON TAB
```

**Comment by kmvanbrunt** [comment]
`cmd2.Cmd` has a method called `_completion_supported()`. Can you check if it's returning `False` and if so, find out which part of the logic is failing?

```Python
def _completion_supported(self) -> bool:
    """Return whether tab completion is supported"""
    return self.use_rawinput and bool(self.completekey) and rl_type != RlType.NONE
```

**Reply by Sripadvallabh** [reply]
```
>>> self._completion_supported()
True
>>>
```

**Reply by Sripadvallabh** [reply]
This i could see from the 'py' shell inside the app

**Reply by Sripadvallabh** [reply]
don't know why \<TAB\> doesn't complete then

**Comment by kmvanbrunt** [comment]
Is `self.completekey` set to `"tab"`?

**Reply by Sripadvallabh** [reply]
```
>>> self.completekey
'tab'
>>>
```

**Reply by Sripadvallabh** [reply]
i don't know why this issue is seen on one of my users; for others after installing readline it worked.

**Comment by Sripadvallabh** [comment]
Let me take the time to say i really really thank you folks for developing such an amazing module & your efforts to maintain it. I find immediate responses to all queries i have & everyone in my team as well the users of my tool appreciate the thoughts gone into this module. Truly Amazing!!!

**Comment by kmvanbrunt** [comment]
I'm out of ideas and unfortunately do not have a Mac to test on. Since it is only happening to one user, I would check their terminal settings to see if they are different from the other users' settings.

**Reply by Sripadvallabh** [reply]
Sure i will follow up with them on this. Thanks for your time
